app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt

```diff
@@ -0,0 +1,332 @@
+            else -> R.id.turboButton
+        }
+
+        return State(
+            focusNode = FocusNode(
+                focusedNavUrlInputNode.viewId,
+                nextFocusUpId,
+                nextFocusDownId),
+            focused = focused)
+    }
+
+    private fun updateReloadButtonFocusTree(
+        focusedReloadButtonNode: FocusNode,
+        sessionState: SessionRepo.State
+    ): State {
+
+        assert(focusedReloadButtonNode.viewId == R.id.navButtonReload)
+
+        val nextFocusLeftId = when {
+            sessionState.forwardEnabled -> R.id.navButtonForward
+            sessionState.backEnabled -> R.id.navButtonBack
+            else -> R.id.navButtonReload
+        }
+
+        return State(
+            focusNode = FocusNode(
+                focusedReloadButtonNode.viewId,
+                nextFocusLeftId = nextFocusLeftId))
+    }
+
+    private fun updateForwardButtonFocusTree(
+        focusedForwardButtonNode: FocusNode,
+        sessionState: SessionRepo.State
+    ): State {
+
+        assert(focusedForwardButtonNode.viewId == R.id.navButtonForward)
+
+        val nextFocusLeftId = when {
+            sessionState.backEnabled -> R.id.navButtonBack
+            else -> R.id.navButtonForward
+        }
+
+        return State(
+            focusNode = FocusNode(
+                focusedForwardButtonNode.viewId,
+                nextFocusLeftId = nextFocusLeftId))
+    }
+
+    private fun updatePocketMegaTileFocusTree(
+        focusedPocketMegatTileNode: FocusNode,
+        pinnedTilesIsEmpty: Boolean,
+        focused: Boolean = true
+    ): State {
+
+        assert(focusedPocketMegatTileNode.viewId == R.id.pocketVideoMegaTileView ||
+            focusedPocketMegatTileNode.viewId == R.id.megaTileTryAgainButton)
+
+        val nextFocusDownId = when {
+            pinnedTilesIsEmpty -> R.id.settingsTileContainer
+            else -> R.id.tileContainer
+        }
+
+        return State(
+            focusNode = FocusNode(
+                focusedPocketMegatTileNode.viewId,
+                nextFocusDownId = nextFocusDownId),
+            focused = focused)
+    }
+
+    /**
+     * Two possible scenarios for losing focus when in overlay:
+     * 1. When all the pinned tiles are removed, tilContainer no longer needs focus
+     * 2. When click on [megaTileTryAgainButton]
+     */
+    private fun handleLostFocusInOverlay(
+        lostFocusNode: FocusNode,
+        sessionState: SessionRepo.State,
+        pinnedTilesIsEmpty: Boolean,
+        pocketState: PocketVideoRepo.FeedState
+    ): State {
+
+        assert(lostFocusNode.viewId == R.id.tileContainer ||
+                lostFocusNode.viewId == R.id.megaTileTryAgainButton)
+
+        val viewId = when (pocketState) {
+            PocketVideoRepo.FeedState.FetchFailed -> R.id.megaTileTryAgainButton
+            PocketVideoRepo.FeedState.Inactive -> R.id.navUrlInput
+            else -> R.id.pocketVideoMegaTileView
```

Comment:
Same as above

> Could we enumerate the other states here, instead of using else? That way if we add any in the future, this will need to be updated before it'll compile.
>
> LoadComplete || Loading || NoAPIKey -> R.id.pocketVideoMegaTileView